### PR TITLE
Settings page

### PR DESCRIPTION
**Pull request by lannaa**

### CODE REVIEW

File: TestsCalculator/SettingsPageTests.cs [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+            //Act
+            settingsPage.CurrencyFormatOptions.SelectByText(currency);
+            settingsPage.SaveBtn.Click();
+            IAlert alert = driver.SwitchTo().Alert();
+            alert.Accept();
+
+            //Assert
+            Assert.IsTrue(calculatorPage.CurrencySymbol.Displayed);
+            // currency symbol is not found 
+            //  Assert.AreEqual(currencySymbol, calculatorPage.CurrencySymbol);
```
Comment (lannaa) [inline TestsCalculator/SettingsPageTests.cs:181]: error "  Expected: \"£\"\n  But was:  <Element (id = d4bd4ce8-4341-4e9e-ad3b-6b7dd23d549f)>\n"

File: TestsCalculator/CalculatorPageTests.cs [hunk 2]
```diff
@@ -195,5 +195,30 @@ public void CheckDefaultStartDate()
             //Assert
             Assert.AreEqual(DateTime.Today.ToString("d/M/yyyy"), calculatorPage.StartDate);
         }
+
+        [Test]
+        public void CheckSettingsBtnDisplayed()
```
Comment (varajan) [inline TestsCalculator/CalculatorPageTests.cs:200]: No need to add tests to verify control exists, if you will click it next. The 'click' action will fail, if the control is not shown.

File: TestsCalculator/CalculatorPageTests.cs [hunk 3]
```diff
@@ -195,5 +195,30 @@ public void CheckDefaultStartDate()
...
+            //Arrange
+            CalculatorPage calculatorPage = new CalculatorPage(driver);
+
+            //Act
+            calculatorPage.SettingsBtn.Click();
+
+            //Assert
+            string actualUrl = driver.Url;
+            string expectedUrl = "http://127.0.0.1:8080/Settings";
+            Assert.AreEqual(expectedUrl, actualUrl);
```
Comment (varajan) [inline TestsCalculator/CalculatorPageTests.cs:221]: You may add a property to a Settings page-object:
```
bool IsPageOpened => driver.Url.Contains("Settings");
```

```suggestion
            settingsPage = new SettingsPage(driver); // add this to Arrange section
            Assert.IsTrue(settingsPage.IsOpened);
```

File: TestsCalculator/Pages/CalculatorPage.cs [hunk 4]
```diff
@@ -25,6 +25,8 @@ public CalculatorPage(IWebDriver driver)
         public string Income => _driver.FindElement(By.Id("income")).GetAttribute("value");
         public string Interest => _driver.FindElement(By.Id("interest")).GetAttribute("value");
         public string EndDate => _driver.FindElement(By.Id("endDate")).GetAttribute("value");
+        public IWebElement CurrencySymbol => _driver.FindElement(By.XPath("(//td [@id=\"currency\"])"));
```
Comment (varajan) [inline TestsCalculator/Pages/CalculatorPage.cs:28]: You may use  single quote symbol inside xpaths. More over, you may make this property 'string', it is good if you don't use the web-element itself, only its text.

```suggestion
        public string CurrencySymbol => _driver.FindElement(By.XPath("(//td [@id='currency'])")).Text;
```

File: TestsCalculator/Pages/CalculatorPage.cs [hunk 5]
```diff
@@ -25,6 +25,8 @@ public CalculatorPage(IWebDriver driver)
...
+        public IWebElement SettingsBtn => _driver.FindElement(By.XPath("//div[text()=\"Settings\"]"));
```
Comment (varajan) [inline TestsCalculator/Pages/CalculatorPage.cs:29]:
```suggestion
        public IWebElement SettingsBtn => _driver.FindElement(By.XPath("//div[text()='Settings']")); 
```

File: TestsCalculator/Pages/SettingsPage.cs [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+        public IWebElement NumberFormat => _driver.FindElement(By.XPath("//th[text()=\"Number format:\"]"));
+        public SelectElement NumberFormatOptions => new SelectElement(_driver.FindElement(By.Id("numberFormat")));
+        public IWebElement Currency => _driver.FindElement(By.XPath("//th[text()=\"Default currency:\"]"));
+        public SelectElement CurrencyFormatOptions => new SelectElement(_driver.FindElement(By.Id("currency")));
+        public IWebElement SaveBtn => _driver.FindElement(By.Id("save"));
+        public IWebElement CancelBtn => _driver.FindElement(By.Id("cancel"));
+        public IWebElement LogoutBtn => _driver.FindElement(By.XPath("//div[text()=\"Logout\"]"));
+    }
+   
+    }
```
Comment (varajan) [inline TestsCalculator/Pages/SettingsPage.cs:26]: Formating:
```suggestion
    }
}
```

File: TestsCalculator/SettingsPageTests.cs [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+            List<string> expectedDateFormat = new List<string> { "dd/MM/yyyy", "dd-MM-yyyy", "MM/dd/yyyy", "MM dd yyyy" };
+
+            //Act
+            SelectElement s = settingsPage.DateFormatOptions;
+
+            IList<string> actualDateFormat = new List<string>();
+            for (int j = 0; j < s.Options.Count; j++)
+            {
+                actualDateFormat.Add(s.Options.ElementAt(j).Text);
+            }
```
Comment (varajan) [inline TestsCalculator/SettingsPageTests.cs:63]: _For_ No. 1
```suggestion
            IList<string> actualDateFormat = new List<string>();
            for (int j = 0; j < settingsPage.DateFormatOptions.Options.Count; j++)
            {
                actualDateFormat.Add(settingsPage.DateFormatOptions.Options.ElementAt(j).Text);
            }
```

File: TestsCalculator/SettingsPageTests.cs [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+            SettingsPage settingsPage = new SettingsPage(driver);
+            List<string> expectedNumberFormat = new List<string> { "123,456,789.00", "123.456.789,00", "123 456 789.00", "123 456 789,00" };
+
+            //Act
+            SelectElement s = settingsPage.NumberFormatOptions;
+            IList<string> actualNumberFormat = new List<string>();
+            for (int j = 0; j < s.Options.Count; j++)
+            {
+                actualNumberFormat.Add(s.Options.ElementAt(j).Text);
+            }
```
Comment (varajan) [inline TestsCalculator/SettingsPageTests.cs:81]: **For** No.2

```suggestion
            IList<string> actualNumberFormat = new List<string>();
            foreach (IWebElement option in settingsPage.NumberFormatOptions.Options)
            {
                actualNumberFormat.Add(option.Text);
            }
```

File: TestsCalculator/SettingsPageTests.cs [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+            SettingsPage settingsPage = new SettingsPage(driver);
+            List<string> expectedCurrencyFormat = new List<string> { "$ - US dollar", "€ - euro", "£ - Great Britain Pound" };
+
+            //Act
+            SelectElement s = settingsPage.CurrencyFormatOptions;
+            IList<string> actualCurrencyFormat = new List<string>();
+            for (int j = 0; j < s.Options.Count; j++)
+            {
+                actualCurrencyFormat.Add(s.Options.ElementAt(j).Text);
+            }
```
Comment (varajan) [inline TestsCalculator/SettingsPageTests.cs:100]: **For** No.3

```suggestion
            // should add LINQ to usings
            IList<string> actualCurrencyFormat = settingsPage.CurrencyFormatOptions.Options.Select(option => option.Text).ToList();
```

File: TestsCalculator/SettingsPageTests.cs [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+            //Arrange
+            SettingsPage settingsPage = new SettingsPage(driver);
+
+            //Act
+            settingsPage.CancelBtn.Click();
+
+            //Assert
+            string actualUrl = driver.Url;
+            string expectedUrl = "http://127.0.0.1:8080/Deposit";
+            Assert.AreEqual(expectedUrl, actualUrl);
```
Comment (varajan) [inline TestsCalculator/SettingsPageTests.cs:118]: I suggest to add some IsOpened property to DepositPage class.

File: TestsCalculator/SettingsPageTests.cs [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+        }
+
+        [TestCase("MM dd yyyy", "M dd yyyy")]
+        [TestCase("MM/dd/yyyy", "M/dd/yyyy")]
+        [TestCase("dd-MM-yyyy", "dd-M-yyyy")]
+        [TestCase("dd/MM/yyyy", "dd/M/yyyy")]
```
Comment (varajan) [inline TestsCalculator/SettingsPageTests.cs:124]: You may use the format twice. To select on Settings page and to convert date to string(format).
```suggestion
        [TestCase("MM dd yyyy")]
        [TestCase("MM/dd/yyyy")]
        [TestCase("dd-MM-yyyy")]
        [TestCase("dd/MM/yyyy")]
```

File: TestsCalculator/SettingsPageTests.cs [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+        {
+            //Arrange
+            SettingsPage settingsPage = new SettingsPage(driver);
+            CalculatorPage calculatorPage = new CalculatorPage(driver);
+
+            //Act
+            settingsPage.DateFormatOptions.SelectByText(dateFormat);
+            settingsPage.SaveBtn.Click();
+            IAlert alert = driver.SwitchTo().Alert();
+            alert.Accept();
```
Comment (varajan) [inline TestsCalculator/SettingsPageTests.cs:135]: No 'driver' is allowed in tests 😈
How to handle it? Create a method Save in Settings page:
```
public void Save()
{
    SaveBtn.Click();
    driver.SwitchTo().Alert().Accept();
}
```

File: TestsCalculator/SettingsPageTests.cs [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+            //Arrange
+            SettingsPage settingsPage = new SettingsPage(driver);
+            CalculatorPage calculatorPage = new CalculatorPage(driver);
+
+            //Act
+            settingsPage.NumberFormatOptions.SelectByText(number);
+            settingsPage.SaveBtn.Click();
+            IAlert alert = driver.SwitchTo().Alert();
+            alert.Accept();
+            //Fill data to verify
```
Comment (varajan) [inline TestsCalculator/SettingsPageTests.cs:156]: I think this comment is not needed.

File: TestsCalculator/SettingsPageTests.cs [hunk 14]
```diff
@@ -0,0 +1,199 @@
...
+            CalculatorPage calculatorPage = new CalculatorPage(driver);
+
+            //Act
+            settingsPage.CurrencyFormatOptions.SelectByText(currency);
+            settingsPage.SaveBtn.Click();
+            IAlert alert = driver.SwitchTo().Alert();
+            alert.Accept();
+
+            //Assert
+            Assert.IsTrue(calculatorPage.CurrencySymbol.Displayed);
```
Comment (varajan) [inline TestsCalculator/SettingsPageTests.cs:179]: Don't assert is displayed, when you will get text from the control on the next step.

File: TestsCalculator/SettingsPageTests.cs [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+            calculatorPage.AmountField.SendKeys("100000");
+
+            // Assert
+            Assert.AreEqual(income, calculatorPage.Income);
+        }
+
+        [TestCase("£ - Great Britain Pound", "£")]
+        [TestCase("€ - euro", "€")]
+        [TestCase("$ - US dollar", "$")]
+        public void SaveCurrencyFormat(string currency, string currencySymbol)
```
Comment (varajan) [inline TestsCalculator/SettingsPageTests.cs:166]: It's a good rule to add 'Test' to the end of the name of each test.

```suggestion
        public void SaveCurrencyFormatTest(string currency, string currencySymbol)
```

File: TestsCalculator/SettingsPageTests.cs [hunk 16]
```diff
@@ -0,0 +1,199 @@
...
+            //Arrange
+            SettingsPage settingsPage = new SettingsPage(driver);
+
+            //Act
+            settingsPage.LogoutBtn.Click();
+
+            //Assert
+            string actualUrl = driver.Url;
+            string expectedUrl = "http://127.0.0.1:8080/";
+            Assert.AreEqual(expectedUrl, actualUrl);
```
Comment (varajan) [inline TestsCalculator/SettingsPageTests.cs:196]:
```suggestion
            Assert.IsTrue(loginPage.IsOpened);
```

File: TestsCalculator/SettingsPageTests.cs [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+        [Test]
+        public void CheckFieldNamesDisplayed()
+        {
+            //Arrange
+            SettingsPage settingsPage = new SettingsPage(driver);
+
+            //Assert
+            Assert.IsTrue(settingsPage.DateFormatText.Displayed);
+            Assert.IsTrue(settingsPage.NumberFormatText.Displayed);
+            Assert.IsTrue(settingsPage.CurrencyFormatText.Displayed);
```
Comment (varajan) [inline TestsCalculator/SettingsPageTests.cs:46]: I think the better solution is to get elements via XPath and control name. Let's discuss it on a call.

File: TestsCalculator/Pages/SettingsPage.cs [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+    {
+        private IWebDriver _driver;
+
+        public SettingsPage(IWebDriver driver)
+        {
+            _driver = driver;
+        }
+
+        public IWebElement DateFormat => _driver.FindElement(By.XPath("//th[text()=\"Date format:\"]"));
+        public SelectElement DateFormatOptions => new SelectElement(_driver.FindElement(By.Id("dateFormat")));
```
Comment (varajan) [inline TestsCalculator/Pages/SettingsPage.cs:16]: I would name it shorter:
```suggestion
        public SelectElement DateFormat => new SelectElement(_driver.FindElement(By.Id("dateFormat")));
```

File: TestsCalculator/Pages/SettingsPage.cs [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+        public IWebElement NumberFormat => _driver.FindElement(By.XPath("//th[text()=\"Number format:\"]"));
+        public SelectElement NumberFormatOptions => new SelectElement(_driver.FindElement(By.Id("numberFormat")));
```
Comment (varajan) [inline TestsCalculator/Pages/SettingsPage.cs:18]:
```suggestion
        public SelectElement NumberFormat => new SelectElement(_driver.FindElement(By.Id("numberFormat")));
```

File: TestsCalculator/Pages/SettingsPage.cs [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+        public IWebElement Currency => _driver.FindElement(By.XPath("//th[text()=\"Default currency:\"]"));
+        public SelectElement CurrencyFormatOptions => new SelectElement(_driver.FindElement(By.Id("currency")));
```
Comment (varajan) [inline TestsCalculator/Pages/SettingsPage.cs:20]:
```suggestion
        public SelectElement Currency => new SelectElement(_driver.FindElement(By.Id("currency")));
```

**Review by varajan — COMMENTED** [review]
Good work!

File: TestsCalculator/SettingsPageTests.cs (hunk 12 above)
Comment (lannaa) [reply TestsCalculator/SettingsPageTests.cs:135]: corrected

File: TestsCalculator/SettingsPageTests.cs (hunk 11 above)
Comment (lannaa) [reply TestsCalculator/SettingsPageTests.cs:124]: corrected

File: TestsCalculator/Pages/CalculatorPage.cs [hunk 21]
```diff
@@ -49,17 +46,9 @@ public string StartDate
         }
         public void ClickCalculateBtn()
```
Comment (varajan) [inline TestsCalculator/Pages/CalculatorPage.cs:47]: I think the method should be named 'Save'

File: TestsCalculator/Pages/CalculatorPage.cs [hunk 22]
```diff
@@ -49,17 +46,9 @@ public string StartDate
...
         {
-           var wait = new WebDriverWait(_driver, TimeSpan.FromSeconds(10));
-           // WebDriverWait wait = new(TimeSpan.FromSeconds(10));
+            var wait = new WebDriverWait(_driver, TimeSpan.FromSeconds(10));
             wait.Until(SeleniumExtras.WaitHelpers.ExpectedConditions.ElementToBeClickable(CalculateBtn));
```
Comment (varajan) [inline TestsCalculator/Pages/CalculatorPage.cs:50]: It is strange, but it works. As for me it should click after wait.

File: TestsCalculator/Pages/SettingsPage.cs [hunk 23]
```diff
@@ -22,13 +21,11 @@ public SettingsPage(IWebDriver driver) : base(driver)
         public IWebElement CancelBtn => _driver.FindElement(By.Id("cancel"));
         public IWebElement LogoutBtn => _driver.FindElement(By.XPath("//div[text()='Logout']"));
 
-        public void Save()
+        public void ClickSaveBtn()
```
Comment (varajan) [inline TestsCalculator/Pages/SettingsPage.cs:24]: Same here, I think 'Save' title is better.

**Review by varajan — COMMENTED** [review]
Great work, you did it! The tests passed. I have two minor comments. You should decide fix them or not. Do how you think it is better and resolve these comments, notify me and I'll give "approve". Let's merge it!

**Review by varajan — APPROVED** [review]
Please merge.

**Pull request merged**